File: java/demo/CrossCall.java
package demo;

import com.apihub.ApiHub;
import com.apihub.DataHandle;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * CrossCall – 并发客户端（Consumer）。
 * 功能：连接到 {@code ipc:cross}，在独立线程中交替调用 {@code 'add'} 和 {@code 'inv_seri'}，
 * 持续 10 秒后自动退出。可同时启动多个实例以模拟负载。
 * 
 * <p>与 Pascal cross_call 中 Do_Compute 的逻辑完全等价。</p>
 * 
 * @see CrossService
 * @see CrossNode
 */
public class CrossCall {

    /** 运行时长（毫秒），与 Pascal 中 10 * 1000 一致 */
    private static final long RUN_DURATION_MS = 10_000L;

    /**
     * 封装 'add' 远程调用。
     * 与 Pascal cross_call 中 add__ 函数完全等价。
     * 
     * @param a 第一个加数
     * @param b 第二个加数
     * @return 计算结果，若超时或失败返回 0
     */
    private static int add__(int a, int b) {
        try (DataHandle send = new DataHandle("add")) {
            send.writeInt32(a);
            send.writeInt32(b);

            try (DataHandle result = ApiHub.call("demo", send, 1000)) {
                if (result.getSize() == 0) {
                    System.err.printf("[Call] add(%d, %d) 超时或失败%n", a, b);
                    return 0;
                }
                return result.readInt32();
            }
        } catch (Exception e) {
            System.err.println("[Call] add 异常: " + e.getMessage());
            return 0;
        }
    }

    /**
     * 封装 'inv_seri' 远程调用。
     * 数据布局与 Pascal cross_call 中 inv_seri_ 函数完全等价。
     * 
     * @return 调试字符串，包含发送和接收的数据对比
     */
    private static String invSeri__() {
        // 固定测试数据（与 Pascal 完全一致）
        int b = 200;
        int w = 0x10;
        long c = 0x2F;
        long u64 = 0x3F;
        String s = "hello world";
        float f = 3.14f;

        try (DataHandle send = new DataHandle("inv_seri")) {
            // ---------- 按顺序写入（与 Pascal 完全一致） ----------
            send.writeUInt8(b);
            send.writeUInt16(w);
            send.writeUInt32(c);
            send.writeUInt64(u64);
            send.writeStringNullTerminated(s);
            send.writeSingle(f);

            try (DataHandle result = ApiHub.call("demo", send, 1000)) {
                if (result.getSize() == 0) {
                    return "inv_seri 超时或失败";
                }

                // ---------- 按反向顺序读取（与 Pascal 完全一致） ----------
                float fRet = result.readSingle();
                String sRet = result.readStringNullTerminated();
                long u64Ret = result.readUInt64();
                long cRet = result.readUInt32();
                int wRet = result.readUInt16();
                int bRet = result.readUInt8();

                return String.format(
                        "接收数据序 [%d, %d, %d, %d, \"%s\", %.2f] = 发送数据序 [%.2f, \"%s\", %d, %d, %d, %d]",
                        bRet, wRet, cRet, u64Ret, sRet, fRet,
                        fRet, sRet, u64Ret, cRet, wRet, bRet
                );
            }
        } catch (Exception e) {
            return "inv_seri 异常: " + e.getMessage();
        }
    }

    /**
     * 工作线程的主循环，与 Pascal cross_call 中 Do_Compute 完全等价。
     * 
     * @param running 运行标志，用于控制循环退出
     * @param latch   CountDownLatch，在循环结束后 countDown()
     */
    private static void doCompute(AtomicBoolean running, CountDownLatch latch) {
        long startTime = System.currentTimeMillis();
        try {
            while (running.get() && (System.currentTimeMillis() - startTime) < RUN_DURATION_MS) {
                // 与 Pascal 中 TMT19937.Rand32 mod 2 等价：随机选择调用 add 或 inv_seri
                if (ThreadLocalRandom.current().nextBoolean()) {
                    int a = ThreadLocalRandom.current().nextInt(1, Integer.MAX_VALUE);
                    int b = ThreadLocalRandom.current().nextInt(1, Integer.MAX_VALUE);
                    int result = add__(a, b);
                    if (result != 0) {
                        long remaining = RUN_DURATION_MS - (System.currentTimeMillis() - startTime);
                        System.out.printf("[Call] 计算 \"a(%d)+b(%d)\" = 计算结果 %d (%.2f秒以后退出)%n",
                                a, b, result, remaining / 1000.0);
                    }
                } else {
                    String status = invSeri__();
                    long remaining = RUN_DURATION_MS - (System.currentTimeMillis() - startTime);
                    System.out.printf("[Call] %s (%.2f秒退出)%n", status, remaining / 1000.0);
                }

                // 与 Pascal 中 TCompute.Sleep(1) 等价：避免 CPU 空转
                Thread.sleep(1);
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            System.err.println("[Call] 工作线程被中断");
        } finally {
            latch.countDown();
        }
    }

    public static void main(String[] args) throws InterruptedException {
        System.out.println("=== CrossCall (Java) – Concurrent Client ===");

        // ---------- 1. 连接服务（纯消费，不暴露 API） ----------
        ApiHub.resetPrepare();
        ApiHub.prepareClient("ipc:cross", null);

        if (!ApiHub.prepareDone()) {
            System.err.println("[ERROR] prepareDone() failed. Check console output.");
            return;
        }
        System.out.println("[OK] Connected to ipc:cross");

        // ---------- 2. 启动工作线程（与 Pascal TCompute.RunC_NP 等价） ----------
        AtomicBoolean running = new AtomicBoolean(true);
        CountDownLatch latch = new CountDownLatch(1);
        ExecutorService executor = Executors.newSingleThreadExecutor();
        executor.submit(() -> doCompute(running, latch));

        System.out.printf("[INFO] 仿真计算启动（可多开），持续 %d 秒...%n", RUN_DURATION_MS / 1000);

        // ---------- 3. 等待工作线程结束（与 Pascal while thread_running do Sleep(100) 等价） ----------
        // 等待 10 秒后，设置停止标志，并等待线程真正退出
        Thread.sleep(RUN_DURATION_MS + 500); // 多等 500ms 确保最后一次循环完成
        running.set(false);
        latch.await(5, TimeUnit.SECONDS);

        // ---------- 4. 清理资源 ----------
        executor.shutdownNow();
        if (!executor.awaitTermination(2, TimeUnit.SECONDS)) {
            System.err.println("[WARN] 工作线程未能及时终止，强制关闭");
        }

        System.out.println("[OK] 计算完成，清理线程中.");

        ApiHub.exitMainThread();
        ApiHub.shutdown();
        System.out.println("[OK] Client shutdown complete.");
    }
}

File: rust/examples/cross_call.rs
// examples/cross_call.rs
use api_hub_rust::*;
use std::time::{Duration, Instant};
use rand::Rng;

// ========== 远程调用封装（原始二进制） ==========

fn add__(a: i32, b: i32) -> i32 {
    let mut param = match DataHandle::new("add") {
        Ok(h) => h,
        Err(_) => return 0,
    };
    if param.write_int32(a).is_err() || param.write_int32(b).is_err() {
        return 0;
    }
    for attempt in 0..3 {
        let res = match call("demo", param.as_raw(), 15000) {
            Ok(r) => r,
            Err(e) => {
                eprintln!("[add] call 错误 (尝试 {}): {:?}", attempt + 1, e);
                continue;
            }
        };
        let size = get_size(res);
        eprintln!("[add] res ptr: {:p}, size: {} (尝试 {})", res, size, attempt + 1);
        if size > 0 {
            // The returned handle is owned – use from_owned_raw.
            let mut resp = unsafe { DataHandle::from_owned_raw(res) };
            return resp.read_int32().unwrap_or(0);
        }
        // If size is 0, we still need to free the handle (it's owned).
        unsafe { DataHandle::from_owned_raw(res) };
        std::thread::sleep(Duration::from_millis(100));
    }
    0
}

fn inv_seri__() -> String {
    let mut param = match DataHandle::new("inv_seri") {
        Ok(h) => h,
        Err(_) => return "inv_seri 创建句柄失败".to_string(),
    };

    let b: u8 = 200;
    let w: u16 = 0x10;
    let c: u32 = 0x2F;
    let u64: u64 = 0x3F;
    let s = "hello world";
    let f: f32 = 3.14;

    if param.write_uint8(b).is_err()
        || param.write_uint16(w).is_err()
        || param.write_uint32(c).is_err()
        || param.write_uint64(u64).is_err()
        || param.write_string_null_terminated(s).is_err()
        || param.write_single(f).is_err()
    {
        return "inv_seri 写入参数失败".to_string();
    }

    for attempt in 0..3 {
        let res = match call("demo", param.as_raw(), 15000) {
            Ok(r) => r,
            Err(e) => {
                eprintln!("[inv_seri] call 错误 (尝试 {}): {:?}", attempt + 1, e);
                continue;
            }
        };
        let size = get_size(res);
        eprintln!("[inv_seri] res ptr: {:p}, size: {} (尝试 {})", res, size, attempt + 1);
        if size > 0 {
            let mut resp = unsafe { DataHandle::from_owned_raw(res) };
            let f_ret = resp.read_single().unwrap_or(0.0);
            let s_ret = resp.read_string_null_terminated().unwrap_or_default();
            let u64_ret = resp.read_uint64().unwrap_or(0);
            let c_ret = resp.read_uint32().unwrap_or(0);
            let w_ret = resp.read_uint16().unwrap_or(0);
            let b_ret = resp.read_uint8().unwrap_or(0);
            return format!(
                "接收数据序 [{}, {}, {}, {}, \"{}\", {:.2}] = 发送数据序 [{:.2}, \"{}\", {}, {}, {}, {}]",
                b_ret, w_ret, c_ret, u64_ret, s_ret, f_ret,
                f_ret, s_ret, u64_ret, c_ret, w_ret, b_ret
            );
        }
        unsafe { DataHandle::from_owned_raw(res) };
        std::thread::sleep(Duration::from_millis(100));
    }
    "inv_seri 超时或失败（重试 3 次）".to_string()
}

fn main() -> Result<()> {
    println!("=== CrossCall (Rust) – Concurrent Client ===");

    let running = std::sync::Arc::new(std::sync::atomic::AtomicBool::new(true));
    let r = running.clone();
    ctrlc::set_handler(move || {
        eprintln!("\n[Call] 收到 Ctrl+C 信号，正在退出...");
        r.store(false, std::sync::atomic::Ordering::SeqCst);
    }).expect("设置 Ctrl+C 处理器失败");

    reset_prepare();
    prepare_client("ipc:cross", std::ptr::null_mut())?;
    prepare_done()?;
    println!("[OK] Connected to IPC ipc:cross");

    let stop = running.clone();
    let handle = std::thread::spawn(move || {
        let start = Instant::now();
        let duration = Duration::from_secs(10);
        let mut rng = rand::thread_rng();

        println!("[Call] 仿真计算启动（可多开），持续 10 秒...");

        while stop.load(std::sync::atomic::Ordering::SeqCst) && start.elapsed() < duration {
            if rng.gen::<bool>() {
                let a = rng.gen_range(1..i32::MAX);
                let b = rng.gen_range(1..i32::MAX);
                let result = add__(a, b);
                let remaining = duration.as_secs_f64() - start.elapsed().as_secs_f64();
                if result != 0 {
                    println!(
                        "[Call] 计算 \"a({})+b({})\" = 计算结果 {} ({:.2}秒以后退出)",
                        a, b, result, remaining
                    );
                } else {
                    println!(
                        "[Call] add({}, {}) 返回 0（重试后仍失败） ({:.2}秒退出)",
                        a, b, remaining
                    );
                }
            } else {
                let status = inv_seri__();
                let remaining = duration.as_secs_f64() - start.elapsed().as_secs_f64();
                println!("[Call] {} ({:.2}秒退出)", status, remaining);
            }
            std::thread::sleep(Duration::from_millis(1));
        }

        println!("[Call] 工作线程结束");
    });

    while running.load(std::sync::atomic::Ordering::SeqCst) {
        if handle.is_finished() {
            break;
        }
        std::thread::sleep(Duration::from_millis(100));
    }

    let _ = handle.join();

    println!("[OK] 计算完成，清理线程中.");

    exit_main_thread();
    shutdown();
    println!("[OK] Client shutdown complete.");
    Ok(())
}